### PR TITLE
Fix docker

### PR DESCRIPTION
**Pull request by shivanth**
https://github.com/quantumlib/Cirq/issues/4134

### CODE REVIEW

**Review by balopat — CHANGES_REQUESTED** [review]
Thank you, I think this looks good. As mentioned in the issue, can you add a test to ensure that it builds the right thing? I'm thinking of something simple, like, after building the image, assert that `docker run cirq-image python -c "import cirq; assert cirq.__version__ is not None"` succeeds or something similar.

**Comment by shivanth** [comment]
I've added have added the test, but unfortunately will need to disable it since the Github actions does not support docker build/run commands. The added test is also a `longrun` test, will increase the pipeline completion time. 
Will add a pytest marker and remove this test from CI executions

**Comment by balopat** [comment]
This is great! We don't need the marker - we should just limit that test to Linux only.

**Comment by shivanth** [comment]
> we should just limit that test to Linux only.
Is there a way to selectively run tests based on os type/version? 
One other way would be to setup a different github action to run docker tests.

**Comment by balopat** [comment]
> > we should just limit that test to Linux only.
> 
> Is there a way to selectively run tests based on os type/version?
> One other way would be to setup a different github action to run docker tests.

I'd pull out the `@only_on_posix` decorator from below to a file called `dev_tools/markers.py`, import it in `dev_tools/__init__.py` and then you can just use it as `@dev_tools.only_on_posix`: 

https://github.com/quantumlib/Cirq/blob/700a6c02efd270bba5969d6df38da741ecb17c09/dev_tools/bash_scripts_test.py#L24

**Comment by balopat** [comment]
> > > we should just limit that test to Linux only.
> > 
> > 
> > Is there a way to selectively run tests based on os type/version?
> > One other way would be to setup a different github action to run docker tests.
> 
> I'd pull out the `@only_on_posix` decorator from below to a file called `dev_tools/markers.py`, import it in `dev_tools/__init__.py` and then you can just use it as `@dev_tools.only_on_posix`:
> 
> https://github.com/quantumlib/Cirq/blob/700a6c02efd270bba5969d6df38da741ecb17c09/dev_tools/bash_scripts_test.py#L24

Hmm...the Mac platform might match posix. We should use [platform](https://stackoverflow.com/questions/1854/python-what-os-am-i-running-on).

Also, as it's just a single test, an even simpler approach is just to have a simple if statement on top of the test which tests for `Linux` and if it's not, then use `pytest.skip()`

File: dev_tools/docker_test.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+def test_docker():
+    if platform.system() != 'Linux':
+        pytest.skip("Unsupported os")
+    root_folder = pathlib.Path(__file__).parent.parent
+    buildResult = subprocess.run(['docker', 'build', '-t', 'cirq_image', '.'], cwd=root_folder)
+    assert buildResult.returncode == 0
+
+    result = subprocess.run(
+        ["docker run cirq_image python -c \"import cirq; assert cirq.__version__ is not None\""],
```
Comment (balopat) [inline dev_tools/docker_test.py:15]:
```suggestion
        ['docker run cirq_image python -c "import cirq; assert cirq.__version__ is not None"'],
```

**Comment by balopat** [comment]
Now, that I'm looking at this Dockerfile, it's a bit of a mess.

We are adding the whole working directory, but not really using it besides for installing the requirements for a development environment. Then we install the latest _stable_ version of cirq, which has nothing to do with the repo context...

We have to decide what we want to do with Docker. 

If we want to publish to Dockerhub (or GCR) our image, then we'll probably want to have the options for pre-release docker images as well as latest stable images, then the `pip install cirq` and `pip install cirq --pre` direction is the correct one. However, this case there is actually no need for development time requirements. I think we should go towards this direction (see https://github.com/quantumlib/Cirq/issues/3346). Which means that we should remove all the dev requirement installation and the addition of the repo files.

If we want to allow people to _develop_ cirq using Docker, then we should set it up differently, and allow for the repo mounted in as the working directory. Then we do need the dev env requirements. I'd argue that there is no true need at the moment for developers to use Docker as the dev environment.

Sorry to realize this here - but then the 4 minute test made me thinking whether we really need this complexity or not on our hands, as it seems no one really uses Docker.

**Comment by shivanth** [comment]
I can came up with a minimal python cirq installed docker image. This should serve as the base image for cirq docker

```python 
FROM python:3.8-slim AS compile-image

# Install dependencies.
# rm -rf /var/lib/apt/lists/* cleans up apt cache. See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/
RUN DEBIAN_FRONTEND=noninteractive apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
     python3-pip \
     locales \
     && rm -rf /var/lib/apt/lists/*

# Configure UTF-8 encoding.
RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && locale-gen
ENV LANG en_US.UTF-8  
ENV LANGUAGE en_US:en  
ENV LC_ALL en_US.UTF-8 

# Make python3 default
RUN rm -f /usr/bin/python \
     && ln -s /usr/bin/python3 /usr/bin/python

# Install cirq
RUN pip3 install cirq

FROM python:3.8-slim AS build-image

# Make sure scripts in .local are usable:
EXPOSE 8888

```


> We have to decide what we want to do with Docker.

1. Docker image  for application that has cirq installed - this is the minimum requirement
2. The above holds for pre release versions as well.
2. Since notebooks are a thing with cirq, it might be worthwhile investing in an image that can support notebooks out of the box.

**Comment by shivanth** [comment]
@balopat requesting review

File: Dockerfile [hunk 2]
```diff
@@ -22,27 +18,27 @@ ENV LC_ALL en_US.UTF-8
...
-COPY requirements.txt .
-COPY ./cirq/contrib/contrib-requirements.txt .
-COPY ./dev_tools/conf/pip-list-dev-tools.txt .
+##cirq pre_release image
+FROM python:3.8-slim AS cirq_pre_release
 
-RUN pip3 install -r requirements.txt -r contrib-requirements.txt -r pip-list-dev-tools.txt
+# Install dependencies.
+# rm -rf /var/lib/apt/lists/* cleans up apt cache. See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/
+RUN DEBIAN_FRONTEND=noninteractive apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
```
Comment (balopat) [inline Dockerfile:28]: you could dedupe these to a `cirq_base` stage and use that as the base layer for the other stages `from cirq_base as cirq_stable` and `from cirq_base as cirq_pre`

**Comment by balopat** [comment]
> I came up with a minimal python cirq installed docker image. This should serve as the base image for cirq docker
> 
> ```python
> FROM python:3.8-slim AS compile-image
> 
> # Install dependencies.
> # rm -rf /var/lib/apt/lists/* cleans up apt cache. See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/
> RUN DEBIAN_FRONTEND=noninteractive apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
>      python3-pip \
>      locales \
>      && rm -rf /var/lib/apt/lists/*
> 
> # Configure UTF-8 encoding.
> RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && locale-gen
> ENV LANG en_US.UTF-8  
> ENV LANGUAGE en_US:en  
> ENV LC_ALL en_US.UTF-8 
> 
> # Make python3 default
> RUN rm -f /usr/bin/python \
>      && ln -s /usr/bin/python3 /usr/bin/python
> 
> # Install cirq
> RUN pip3 install cirq
> ```
> 
> > We have to decide what we want to do with Docker.
> 
> 1. A base docker image that has cirq pre-installed - this is the minimum requirement
> 2. Another image , like the above , for pre-release versions as well.
> 3. Since notebooks are a thing with cirq, it might be worthwhile investing in an image that can support notebooks out of the box.

I like the combined stable / pre multi-stage image.
The notebook idea is valid as well - but then we have to figure out how to copy the right notebooks to the image - some notebooks will only work with pre and won't work with the stable version...I guess the user can figure that out then and use the pre image?

**Review by balopat — APPROVED** [review]
LGTM, thank you, it's much better for now.
Eventually we can evolve this further if there is user feedback.

**Comment by shivanth** [comment]
xref #3736

**Pull request merged**